### PR TITLE
Allow customizating Jdbc PropertyResultSetExtractor - Fixes gh-1952

### PR DESCRIPTION
**Pull request by beatfreaker**
Allow customizating Jdbc PropertyResultSetExtractor - Fixes gh-1952

@ryanjbaxter do you think this is the correct approach?

### CODE REVIEW

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java [hunk 1]
```diff
@@ -57,19 +56,21 @@
...
+	private final PropertiesResultSetExtractor extractor;
 
 	private int order;
 
 	private String sql;
 
 	private boolean failOnError;
 
-	public JdbcEnvironmentRepository(JdbcTemplate jdbc, JdbcEnvironmentProperties properties) {
+	public JdbcEnvironmentRepository(JdbcTemplate jdbc, JdbcEnvironmentProperties properties,
```
Comment (ryanjbaxter) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java:67]: Make this as deprecated and create a new constructor with the addition parameter

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/PropertiesResultSetExtractor.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+package org.springframework.cloud.config.server.environment;
+
+import java.util.Map;
+
+import org.springframework.jdbc.core.ResultSetExtractor;
+
+/**
+ * @author Chintan Radia
+ */
+public interface PropertiesResultSetExtractor extends ResultSetExtractor<Map<String, ?>> {
```
Comment (ryanjbaxter) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/PropertiesResultSetExtractor.java:26]: Not sure I understand the need for this interface

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]
Could you submit this agains the 3.0.x branch?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/PropertiesResultSetExtractor.java (hunk 2 above)
Comment (beatfreaker) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/PropertiesResultSetExtractor.java:26]: was trying to expose an interface which a developer can implement to provide a custom PropertyResultSetExtractor implementation.
Without this, we have to modify the existing class as below.
```java
public static class PropertiesResultSetExtractor implements ResultSetExtractor<Map<String, Object>> {

		@Override
		public Map<String, Object> extractData(ResultSet rs) throws SQLException, DataAccessException {
			Map<String, Object> map = new LinkedHashMap<>();
			while (rs.next()) {
				map.put(rs.getString(1), rs.getString(2));
			}
			return map;
		}
	}
```

and if anyone wants to provide customer implementation, they have to create a child class and provide its bean.
and using AutoConfiguration class we can control If we want to use `PropertiesResultSetExtractor` or `CustomPropertiesResultSetExtractor`

```java
public static class CustomResultSetExtractor extends JdbcEnvironmentRepository.PropertiesResultSetExtractor {
@Override
public Map<String, Object> extractData(ResultSet rs) throws SQLException, DataAccessException {
	Map<String, Object> map = new LinkedHashMap<>();
	while (rs.next()) {
		map.put(rs.getString(1), rs.getInt(2));
	}
	return map;
}
}


@Bean
@ConditionalOnMissingBean(value = JdbcEnvironmentRepository.PropertiesResultSetExtractor.class)
public JdbcEnvironmentRepository.PropertiesResultSetExtractor propertiesResultSetExtractor() {
	return new JdbcEnvironmentRepository.PropertiesResultSetExtractor();
}
```

Is this the better approach?

Comment (ryanjbaxter) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/PropertiesResultSetExtractor.java:26]: You could just make `PropertiesResultSetExctractor` a top level class, that might be easiest, then users can just extend that.  Then use the `@ConditionalOnMissingBean` approach.

**Pull request closed without merging**